diff --git a/Android.mk b/Android.mk
index 5e5da57..3f37941 100644
--- a/Android.mk
+++ b/Android.mk
@@ -16,13 +16,13 @@ VERSION ?= us
 TOUCH_CONTROLS ?= 1
 
 # Disable text-based save-files by default
-TEXTSAVES ?= 0
+TEXTSAVES ?= 1
 
 # Enable extended options menu by default
 EXT_OPTIONS_MENU ?= 1
 
 # Load resources from external files
-EXTERNAL_DATA ?= 0
+EXTERNAL_DATA ?= 1
 BASEDIR ?= res
 
 ifeq ($(VERSION),jp)
@@ -46,7 +46,7 @@ endif
 
 PC_BUILD_DIR := $(LOCAL_PATH)/build/$(VERSION)_pc
 
-SRC_DIRS := src src/engine src/game src/audio src/menu src/buffers actors levels bin bin/$(VERSION) data assets src/pc src/pc/gfx src/pc/audio src/pc/controller src/pc/fs
+SRC_DIRS := src src/engine src/game src/audio src/menu src/buffers actors levels bin bin/$(VERSION) data assets src/pc src/pc/gfx src/pc/audio src/pc/controller src/pc/fs src/sgi src/sgi/utils
 SRC_DIRS := $(addprefix $(LOCAL_PATH)/,$(SRC_DIRS))
 
 GODDARD_SRC_DIRS := src/goddard src/goddard/dynlists
@@ -76,7 +76,7 @@ ULTRA_C_FILES := $(addprefix $(LOCAL_PATH)/lib/src/,$(ULTRA_C_FILES))
 
 GODDARD_C_FILES := $(foreach dir,$(GODDARD_SRC_DIRS),$(wildcard $(dir)/*.c))
 
-GENERATED_C_FILES := $(PC_BUILD_DIR)/assets/mario_anim_data.c $(PC_BUILD_DIR)/assets/demo_data.c \
+GENERATED_C_FILES := $(PC_BUILD_DIR)/assets/mario_anim_data.c $(PC_BUILD_DIR)/assets/luigi_anim_data.c $(PC_BUILD_DIR)/assets/demo_data.c \
   $(addprefix $(PC_BUILD_DIR)/bin/,$(addsuffix _skybox.c,$(notdir $(basename $(wildcard $(LOCAL_PATH)/textures/skyboxes/*.png)))))
 
 LOCAL_SHORT_COMMANDS := true
diff --git a/Makefile b/Makefile
index a6dbdf0..5d8c3e5 100644
--- a/Makefile
+++ b/Makefile
@@ -34,6 +34,7 @@ OSX_BUILD ?= 0
 TARGET_ARCH ?= native
 TARGET_BITS ?= 0
 
+TOUCHCONTROLS ?= 1
 # Disable better camera by default
 BETTERCAMERA ?= 0
 # Disable no drawing distance by default
@@ -541,6 +542,10 @@ else
 endif
 
 # Check for enhancement options
+ifeq ($(TOUCHCONTROLS),1)
+  CC_CHECK += -DTOUCHCONTROLS
+  CFLAGS += -DTOUCHCONTROLS
+endif
 
 # Check for Puppycam option
 ifeq ($(BETTERCAMERA),1)
diff --git a/Makefile.split b/Makefile.split
index 7657f4f..7e7168e 100644
--- a/Makefile.split
+++ b/Makefile.split
@@ -5,7 +5,7 @@
 # obtain a list of segments from the *.c files in bin directory
 SEGMENTS := $(notdir $(basename $(wildcard bin/*.c))) $(addprefix $(VERSION)/,$(notdir $(basename $(wildcard bin/$(VERSION)/*.c)))) $(addsuffix _skybox,$(notdir $(basename $(wildcard textures/skyboxes/*.png))))
 ACTORS := $(filter %/,$(wildcard actors/*/))
-TEXTURE_DIRS := $(addprefix textures/,$(SEGMENTS)) $(ACTORS) textures/crash_screen textures/intro_raw textures/ipl3_raw textures/skybox_tiles
+TEXTURE_DIRS := $(addprefix textures/,$(SEGMENTS)) $(ACTORS) textures/crash_screen textures/intro_raw textures/ipl3_raw textures/skybox_tiles textures/touchcontrols
 
 # NOTE: textures assume naming convention "texture.<encoding>.png" generates "texture.<encoding>"
 
@@ -189,6 +189,7 @@ SKY_FILES                    := $(patsubst %.png,%.inc.c,$(wildcard $(TEXTURE_DI
 MACHINE_FILES                := $(patsubst %.png,%.inc.c,$(wildcard $(TEXTURE_DIR)/machine/*.png))
 MOUNTAIN_FILES               := $(patsubst %.png,%.inc.c,$(wildcard $(TEXTURE_DIR)/mountain/*.png))
 GRASS_FILES                  := $(patsubst %.png,%.inc.c,$(wildcard $(TEXTURE_DIR)/grass/*.png))
+TOUCH_FILES                  := $(patsubst %.png,%.inc.c,$(wildcard $(TEXTURE_DIR)/touchcontrols/*.png))
 
 # Texture Files
 $(BUILD_DIR)/bin/segment2.o: $(addprefix $(BUILD_DIR)/,$(SEGMENT2_FILES))
@@ -206,6 +207,7 @@ $(BUILD_DIR)/bin/sky.o: $(addprefix $(BUILD_DIR)/,$(SKY_FILES))
 $(BUILD_DIR)/bin/machine.o: $(addprefix $(BUILD_DIR)/,$(MACHINE_FILES))
 $(BUILD_DIR)/bin/mountain.o: $(addprefix $(BUILD_DIR)/,$(MOUNTAIN_FILES))
 $(BUILD_DIR)/bin/grass.o: $(addprefix $(BUILD_DIR)/,$(GRASS_FILES))
+$(BUILD_DIR)/src/pc/controller/controller_touchscreen.o: $(addprefix $(BUILD_DIR)/,$(TOUCH_FILES))
 
 # Others
 $(BUILD_DIR)/bin/segment2.elf: SEGMENT_ADDRESS := 0x02000000
diff --git a/src/engine/level_script.c b/src/engine/level_script.c
index 7de998e..568a2a3 100644
--- a/src/engine/level_script.c
+++ b/src/engine/level_script.c
@@ -860,6 +860,10 @@ static void (*LevelScriptJumpTable[])(void) = {
     /*3E*/ level_cmd_cleardemoptr,
 };
 
+#ifdef TOUCH_CONTROLS
+extern void render_touch_controls(void);
+#endif
+
 struct LevelCommand *level_script_execute(struct LevelCommand *cmd) {
     sScriptStatus = SCRIPT_RUNNING;
     sCurrentCmd = cmd;
@@ -871,6 +875,9 @@ struct LevelCommand *level_script_execute(struct LevelCommand *cmd) {
     profiler_log_thread5_time(LEVEL_SCRIPT_EXECUTE);
     init_render_image();
     render_game();
+#ifdef TOUCH_CONTROLS
+    render_touch_controls();
+#endif
     end_master_display_list();
     alloc_display_list(0);
 
diff --git a/src/game/behavior_actions.c b/src/game/behavior_actions.c
index b40408b..5ff7bc5 100644
--- a/src/game/behavior_actions.c
+++ b/src/game/behavior_actions.c
@@ -43,6 +43,7 @@
 #include "spawn_object.h"
 #include "spawn_sound.h"
 #include "thread6.h"
+#include "src/sgi/utils/characters.h"
 
 #define o gCurrentObject
 
